src/shims/unix/fd.rs

```diff
@@ -571,7 +571,10 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 fd.borrow_mut().pread(communicate, &mut bytes, offset, this)
             }
         };
-        if let Ok(Ok(_)) = result {
+        let is_socketpair_fd = fd.borrow().name() == "socketpair";
+        // edge case for socketpair: only the peer socketpair needs to be updated and it is
+        // currently done under FileDescription::read of socketpair.
+        if is_socketpair_fd && let Ok(Ok(_)) = result {
```

Comment:
We should definintely not special case on the name of the fd.

Is it actually a problem to check and update the readiness of socketpair? Won't it just recreate the current state in the ready list?

Reply:
Let's say we have a socketpair with file descriptor value ``fd[0]`` and ``fd[1]``. If we write to ``fd[0]``, we should only receive notification for ``fd[1]`` instead of ``fd[0]``. The problem of not skipping the check mentioned is we will receive notification for both ``fd[0]`` and ``fd[1]``, which is more notification than what should be received in real libc.

Reply:
Maybe we should return `InterpResult<'tcx, (io::Result<usize>, EpollFlags)>` or sth similar, so that every read/write can decide whether they should actually do a flag update

`EpollFlags` can either be an enum with variants stating "yes/no" for doing an update (so a strongly typed bool) or we could literally return the flags, removing the need for `get_epoll_flags` entirely (from the trait, it's still useful as an inherent method on the file descriptors that need it). Though now that I think of it, the latter probably makes the socketpair "notify peer fd" harder, and stops protecting us from putting non-epoll-supported fds into the epoll interest list. So let's just do the in-band signalling via write/read return type.

We just need to decide whether to use

```
InterpResult<'tcx, (io::Result<usize>, EpollFlags)>
```

or

```
InterpResult<'tcx, io::Result<(usize, EpollFlags)>>
```

The latter reflects the current situation where errors in r/w never update flags

Reply:
Wait why is this signaled via write/read return types? That seems odd to me. The `write`/`read` shoms shouldn't have to worry about epoll things. that seems like a layering violation to me.

Reply:
I thought so, too, but apparently epoll is weirder than that and notifies a lot more than just "you can read now"

Reply:
So let's look at this from scratch again. Socketpair works just fine. the question is what eventfd needs to do on r/w. Maybe it needs a weak fd ref to itself?

Reply:
> So let's look at this from scratch again. Socketpair works just fine. the question is what eventfd needs to do on r/w. Maybe it needs a weak fd ref to itself?

Yup, this will make it possible to invoke ``check_and_update_readiness`` in socket.rs ``FileDescription::read/write``. 

Another way that can make eventfd works is having an ``id`` field in eventfd struct`, then we move ``check_and_update_readiness`` as part of ``FileDescription`` trait.

Reply:
Ah, interesting.

Okay then, what about the following:
- `FileDescriptor` gains a method like `describe_readiness` or so that returns all readiness flags: `fn epoll_readiness(...) -> InterpResult<EpollEvents>` or so. (Seems like the event set is represented as `events: u32` in `EpollInterest`, but would probably be good to have at least a type alias for clarity.)
- Socketpair, when doing a write to a previously empty buffer (or when doing any write? needs to be tested), calls a method like `epoll_notify_for_fd` that takes a `FileDescriptionRef`. It can then determine whether anyone cares about that FD, and it can call `epoll_readiness` to determine what to notify about.